internal/docstore/firedocstore/fs.go

```diff
@@ -33,10 +33,25 @@ import (
 )
 
 type collection struct {
-	client    *vkit.Client
-	dbPath    string // e.g. "projects/P/databases/(default)"
-	collPath  string // e.g. "projects/P/databases/(default)/documents/States/Wisconsin/cities"
-	nameField string
+	opts     Options
+	client   *vkit.Client
+	dbPath   string // e.g. "projects/P/databases/(default)"
+	collPath string // e.g. "projects/P/databases/(default)/documents/States/Wisconsin/cities"
+
+}
+
+// Options sets options for constructing a *docstore.Collection backed by Google Firestore.
+type Options struct {
+	// The document field to use for the document name.
+	// Exactly one of NameField or NameFunc must be set.
+	NameField string
```

Comment:
It's a bit unfortunate that one of these is required, but they are hidden away in `Options`.

Also, IIUC, having a dual key now results in (likely) storing them twice (once, combined, as the name, and once separately in the doc), unlike a single key. Presumably this is because you can't recover the two fields from the name?

WDYT about something like:

```
type Namer interface {
  // Name returns a unique name for the document, along with a slice of
  // fields to delete from the document itself (because they are captured by
  // the name).
  Name(doc docstore.Document /* or should this be pdoc? */) (string, []string)
  
  // FromName processes a document name into the fields it came from.
  FromName(name string) map[string]string
}
```

We could have simple helpers to create single field and dual field implementations of `Namer`. For example, the simple case of a name based on a single field would return the value of the field plus the field name from `Name`, and a `map` from the field name to the passed-in name for `FromName`. A two-field `Namer` would concatenate the two fields somehow in `Name`, and return that plus the two field names; in `FromName`, it would reverse the concatenation operation on the passed-in name and use it to create a map from the two field names to the two split values.

The `OpenDocument` constructor would take an instance of `Namer`:

```
OpenCollection(client *vkit.Client, projectID, collPath string, namer Namer, opts *Options)
```

Alternatively, we could have separate constructors, e.g. `OpenCollection` would take a single field name and create a `Namer` internally, and have a separate `OpenCollectionWithNamer` or whatever.

Reply:
Regarding Options: I split into two separate OpenXXX functions.

I don't think we can remove the primary-key fields from the document because they're needed for queries.

Reply:
Why is it OK to remove the primary key field when there's a single one? Are we clever enough to redirect all query filters on that field to be on the document name?

Reply:
That's a good point. I forgot about handling queries on the name. I'll have to see how that works in firestore. I filed #1886.

The problem with multiple primary key fields is that the driver couldn't know how they're combined into the key. A collection with `Game` and `User` fields may construct the key as `Game,User` or `Game|User` or `User/Game`. How could the driver translate the filter `User = "bob"` into a query on the key?